airlock/types.py

```diff
@@ -12,3 +18,70 @@
     class UrlPath(PurePosixPath): ...
 else:
     UrlPath = PurePosixPath
+
+
+@dataclass
+class FileMetadata:
+    """Represents the base properties of file metadata.
+
+    Often these are in the manifest.json, so can be read from there. But
+    sometimes they are not, so need reading from the filesystem.
+    """
+
+    size: int | None
+    timestamp: int | None
+    _content_hash: str | None = None
+    path: Path | None = None
```

Comment:
It'd be helpful to understand when these values can be null or not.  If they're either all set or all null, that suggests we've not got the design quite right here.

Reply:
Ah -- so a directory can have a `FileMetadata`, whose values are `None`.  This doesn't feel quite right.

Reply:
Could we have a `MetadataInterface` or whatever you'd call it in Python, with `FileMetadata` and `DirMetadata` implementations?

Reply:
We could, but I'm not sure what problem that type would solve? 

This was actually the simplest implementation I could got to. I spent some time in the weeds trying to get the typing to play nice here.

For context, there are two consumers of the FileMetadata interface - we use it in regular logic code, where we can could easily just return a None instead of a FileMetadata in the directory case.

But the other consumer is in the templates, where its much trickier to handle a None value that we need to access attributes of if its not None. Perhaps we can use some template filters to do this? As so make `def get_file_metadata(...) -> FileMetadata | None`, and not need the the FileMetadata.empty constructor and None-able fields?

Reply:
I had a quick look at this, and I can't see a way to do it that doesn't give template variable errors. I think this is a consequence of us not letting template variables error.

We could have `size=-1` instead of None, but that's just as bad.

The default template filter doesn't work on errors, on truthy or none.

In jinja, you can do `x.y|default('foo')` and if x is None, it will work.

Reply:
I have found a fix for the awkward type.

It means more boiler plate properties on PathItem, but that is standard fair for django django templates I think.